File: python-assignments/Array/array-list.py
MAX_SIZE = 100

class ArrayList:
    def __init__(self):
        self.data = [0] * MAX_SIZE
        self.size = 0

# insert value at index
def insert_at(lst, index, value):
    if lst.size == MAX_SIZE:
        print("List is full!!")
        return
    if index < 0 or index > lst.size:
        print("Invalid index")
        return
    # shift right
    for i in range(lst.size, index, -1):
        lst.data[i] = lst.data[i-1]
    lst.data[index] = value
    lst.size += 1

# delete at index
def delete_at(lst, index):
    if lst.size == 0:
        print("List is empty!")
        return
    if index < 0 or index >= lst.size:
        print("Invalid index")
        return
    # shift left
    for i in range(index, lst.size - 1):
        lst.data[i] = lst.data[i+1]
    lst.size -= 1

# print list
def traverse(lst):
    if lst.size == 0:
        print("List is empty")
        return
    for i in range(lst.size):
        print("The data at", i, "is", lst.data[i])

# main
if __name__ == "__main__":
    lst = ArrayList()
    insert_at(lst, 0, 12)
    insert_at(lst, 0, 20)
    insert_at(lst, 0, 90)
    insert_at(lst, 0, 120)
    insert_at(lst, 4, 1000)

    traverse(lst)
    print("Size after insertions:", lst.size)

    delete_at(lst, 0)
    delete_at(lst, 3)

    print("Array after deletion")
    traverse(lst)
    print("Size after two deletions:", lst.size)

File: pascal-assignments/Array/Array-list.pas
program ArrayListDemo;

const
  MAX_SIZE = 100;

var
  data : array[0..MAX_SIZE-1] of integer;
  size : integer;
  i    : integer;

{ insert value at index }
procedure InsertAt(index, value : integer);
var
  i : integer;
begin
  if size = MAX_SIZE then
  begin
    writeln('List is full');
    exit;
  end;

  if (index < 0) or (index > size) then
  begin
    writeln('Invalid index');
    exit;
  end;

  { shift elements to the right }
  for i := size downto index + 1 do
    data[i] := data[i-1];

  data[index] := value;
  size := size + 1;
end;

{ delete value at index }
procedure DeleteAt(index : integer);
var
  i : integer;
begin
  if size = 0 then
  begin
    writeln('List is empty');
    exit;
  end;

  if (index < 0) or (index >= size) then
  begin
    writeln('Invalid index');
    exit;
  end;

  { shift elements to the left }
  for i := index to size - 2 do
    data[i] := data[i+1];

  size := size - 1;
end;

{ print list }
procedure Traverse;
var
  i : integer;
begin
  if size = 0 then
  begin
    writeln('List is empty');
    exit;
  end;

  for i := 0 to size - 1 do
    writeln('The data at ', i, ' is ', data[i]);
end;

begin
  { start }
  size := 0;

  InsertAt(0, 12);
  InsertAt(0, 20);
  InsertAt(0, 90);
  InsertAt(0, 120);
  InsertAt(4, 1000);

  Traverse;
  writeln('Size after insertions: ', size);

  DeleteAt(0);
  DeleteAt(3);

  writeln('Array after deletion');
  Traverse;
  writeln('Size after two deletions: ', size);
end.
